lib/puppet/pops/loader/task_instantiator.rb

```diff
@@ -1,97 +1,43 @@
-          path = executables.find { |real_impl| File.basename(real_impl) == impl['name'] }
-          if path
-            { "name" => impl['name'], "requirements" => impl.fetch('requirements', []), "path" => path }
-          else
-            raise ArgumentError, _("Task metadata for task %{name} specifies missing implementation %{implementation}") %
-              { name: name, implementation: impl['name'] }
-          end
-        end
-      else
-        # If 'implementations' is the wrong type, we just pass it through and
-        # let the task type definition reject it.
-        metadata['implementations']
-      end
-    # If implementations isn't defined, then we use executables matching the
-    # task name, and only one may exist.
-    else
-      implementations = executables.select { |impl| File.basename(impl, '.*') == basename }
-      if implementations.empty?
-        raise ArgumentError, _('No source besides task metadata was found in directory %{directory} for task %{name}') %
-          { name: name, directory: directory }
-      elsif implementations.length > 1
-        raise ArgumentError, _("Multiple executables were found in directory %{directory} for task %{name}; define 'implementations' in metadata to differentiate between them") %
-          { name: name, directory: implementations[0] }
-      end
-
-      [{ "name" => File.basename(implementations.first), "path" => implementations.first, "requirements" => [] }]
-    end
-  end
-
   def self.create(loader, typed_name, source_refs)
     name = typed_name.name
     basename = typed_name.name_parts[1] || 'init'
     dirname = File.dirname(source_refs[0])
     metadata_files, executables = source_refs.partition { |source_ref| source_ref.end_with?('.json') }
     metadata_file = metadata_files.find { |source_ref| File.basename(source_ref, '.json') == basename }
 
-    metadata = load_metadata(loader, metadata_file)
+    metadata = Puppet::Module::Task.read_metadata(metadata_file) || {}
 
-    implementation_metadata = validate_implementations(typed_name, dirname, metadata, executables)
+    implementations = Puppet::Module::Task.find_implementations(name, dirname, metadata, executables)
+    files = Puppet::Module::Task.find_files(metadata)
 
-    arguments = {
-      'name' => name,
-      'implementations' => implementation_metadata
-    }
+    task = { 'name' => name, 'metadata' => metadata, 'files' => implementations + files }
 
     begin
-      metadata.each_pair do |key, value|
-        if %w[parameters output].include?(key)
-          ps = {}
-          value.each_pair do |k, v|
-            pd = v.dup
-            t = v['type']
-            pd['type'] = t.nil? ? Types::TypeFactory.data : Types::TypeParser.singleton.parse(t)
-            ps[k] = pd
-          end
-          value = ps
-        end
-        arguments[key] = value unless arguments.key?(key)
-      end
+      metadata['parameters'] = convert_types(metadata['parameters'])
```

Comment:
do we get value from converting the types or should we try to switch this out too?